### PR TITLE
Improve snapshot actions across devices, instances and applications

### LINKED ISSUE #3815: Add appropriate Snapshot actions to all snapshot table views
### Description

We currently list snapshots in a few different views. We are lacking consistency on the actions available against individual snapshots in each of those places.

Here's what actions we provide currently in the respective views:

 - `instance/:id/snapshots`
    - `Rollback`
    - `Download Snapshot`
    - `Download package.json`
    - `Set as device target`
    - `Delete Snapshot`
 - `device/:id/snapshots`
    -  `Delete Snapshot`
    - A 'deploy snapshot' button appears in the row when the mouse hovers over it instead of using the action menu
- `application/:id/snapshots`
    - No action menu

Proposed set of changes:

 - https://github.com/FlowFuse/flowfuse/issues/3629 already proposes renaming 'Rollback' - should also consider the two different UXes we have for that action
 - Add 'delete snapshot' to Application view
 - Add Download* to Application/device view - don't immediately see why that would be restricted to just instances

Goal is very much a quick iteration to get these things inline.

### Epic/Story

_No response_

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #3815
closes https://github.com/FlowFuse/flowfuse/issues/3629

## Description

* Squares up the kebab menus of snapshot views (see caveats)
  * including renaming rollback to "Deploy Snapshot" (see https://github.com/FlowFuse/flowfuse/issues/3629)
* Provides missing feature Download Snapshot and Download Package.json for devices
   * This was dependant on prior works https://github.com/FlowFuse/flowfuse/pull/3833

### Caveats
Some menu items do not make a sense on other views and thus not all menus are the same. This was discussed and agreed in the [issue comment](https://github.com/FlowFuse/flowfuse/issues/3815#issuecomment-2117831951) 

### TODO: 
- [x] tests - done
- [x] check see if there is anywhere docs should be updated - done
- [x] changelog entry : https://github.com/FlowFuse/website/issues/2079

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Review by cstns — COMMENTED** [review]
partial review, unable to leave more comments for it due to `This diff is outdated, please refresh and try again.`

File: frontend/src/pages/device/Snapshots/index.vue [hunk 1]
```diff
@@ -265,7 +255,25 @@ export default {
         onSnapshotCancel () {
             this.busyMakingSnapshot = false
         },
-
+        async downloadSnapshotPackage (snapshot) {
+            const ss = await SnapshotApi.getFullSnapshot(snapshot.id)
+            const owner = snapshot.device || snapshot.project
```
Comment (cstns) [inline frontend/src/pages/device/Snapshots/index.vue:260]: This data manipulation/transformation seems to be similar to one done before in the applucation/Snapshots.vue. If moving the transformer in the BE is not a viable option, using a common snapsot transformer/mapper would be the way to do it

File: frontend/src/pages/device/Snapshots/index.vue [hunk 2]
```diff
@@ -334,6 +342,26 @@ export default {
...
+        },
+        showDownloadSnapshotDialog (snapshot) {
+            this.$refs.snapshotExportDialog.show(snapshot)
+        },
+        showDeploySnapshotDialog (snapshot) {
+            this.deploySnapshot(snapshot.id)
+        },
+
+        // enable/disable snapshot actions
+        canDeploy (_row) {
```
Comment (cstns) [inline frontend/src/pages/device/Snapshots/index.vue:354]: This methods would be better suite d in computed props which would also sync the state of the button automatically when rows would change. Using methods to determine state changes could cause unexpected behavior (elements not updating after state changes) in the ui

File: frontend/src/pages/instance/Snapshots/index.vue [hunk 3]
```diff
@@ -259,13 +256,15 @@ export default {
...
         },
         async downloadSnapshotPackage (snapshot) {
-            const ss = await SnapshotApi.getSnapshot(this.instance.id, snapshot.id)
+            const ss = await SnapshotsApi.getFullSnapshot(snapshot.id)
+            const owner = snapshot.device || snapshot.project
+            const ownerType = snapshot.device ? 'device' : 'instance'
             const packageJSON = {
-                name: this.instance.safeName,
-                description: `${snapshot.name} - ${snapshot.description}`,
+                name: `${owner.safeName || owner.name}`.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
```
Comment (cstns) [inline frontend/src/pages/instance/Snapshots/index.vue:263]: The `.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase()` could be extracted as a separate composable into the string composable `frontend/src/composables/String.js:5` and reused

File: frontend/src/pages/instance/Snapshots/index.vue [hunk 4]
```diff
@@ -259,13 +256,15 @@ export default {
             this.$emit('instance-updated')
         },
         async downloadSnapshotPackage (snapshot) {
-            const ss = await SnapshotApi.getSnapshot(this.instance.id, snapshot.id)
+            const ss = await SnapshotsApi.getFullSnapshot(snapshot.id)
+            const owner = snapshot.device || snapshot.project
+            const ownerType = snapshot.device ? 'device' : 'instance'
```
Comment (cstns) [inline frontend/src/pages/instance/Snapshots/index.vue:261]: This transformation/mapping seems similar to the ones in the previous components. Would it be possible to extract the entire download method into a composable/mixin?

File: frontend/src/pages/instance/Snapshots/index.vue [hunk 5]
```diff
@@ -259,13 +256,15 @@ export default {
...
-                name: this.instance.safeName,
-                description: `${snapshot.name} - ${snapshot.description}`,
+                name: `${owner.safeName || owner.name}`.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
+                description: `${ownerType} snapshot, ${snapshot.name} - ${snapshot.description}`,
                 private: true,
                 version: '0.0.0-' + snapshot.id,
-                dependencies: ss.modules
+                dependencies: ss.settings?.modules || {}
             }
             const element = document.createElement('a')
```
Comment (cstns) [inline frontend/src/pages/instance/Snapshots/index.vue:269]: We could use the same download snapshot composable trigger

**Review by cstns — COMMENTED** [review]
haven't went over the cypress tests yet, would love to have a chat before starting to make changes

File: frontend/src/pages/instance/Snapshots/index.vue (hunk 5 above)
Comment (Steve-Mcl) [reply frontend/src/pages/instance/Snapshots/index.vue]: resolved in [15001ab](https://github.com/FlowFuse/flowfuse/pull/3885/commits/15001ab22c59d4581f2116b97633c5c418756099)

File: frontend/src/pages/instance/Snapshots/index.vue (hunk 3 above)
Comment (Steve-Mcl) [reply frontend/src/pages/instance/Snapshots/index.vue:252]: not in scope at this time (as discussed)

File: frontend/src/pages/device/Snapshots/index.vue (hunk 2 above)
Comment (Steve-Mcl) [reply frontend/src/pages/device/Snapshots/index.vue:337]: Discussed & I understand the better way now but not in scope at this time

Once Serban has implemented the first one and we establish a pattern, these will be retro'd upon working in the file

File: frontend/src/pages/device/Snapshots/index.vue (hunk 1 above)
Comment (Steve-Mcl) [reply frontend/src/pages/device/Snapshots/index.vue]: resolved in [15001ab](https://github.com/FlowFuse/flowfuse/pull/3885/commits/15001ab22c59d4581f2116b97633c5c418756099)

File: frontend/src/pages/device/Snapshots/index.vue [hunk 6]
```diff
@@ -265,7 +255,25 @@ export default {
...
+            const owner = snapshot.device || snapshot.project
+            const ownerType = snapshot.device ? 'device' : 'instance'
+            const packageJSON = {
+                name: `${owner.safeName || owner.name}`.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
+                description: `${ownerType} snapshot, ${snapshot.name} - ${snapshot.description}`,
+                private: true,
+                version: '0.0.0-' + snapshot.id,
+                dependencies: ss.settings?.modules || {}
+            }
+            const element = document.createElement('a')
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/device/Snapshots/index.vue]: resolved in [15001ab](https://github.com/FlowFuse/flowfuse/pull/3885/commits/15001ab22c59d4581f2116b97633c5c418756099)

File: frontend/src/pages/application/Snapshots.vue [hunk 7]
```diff
@@ -135,6 +156,61 @@ export default {
...
+                kind: 'danger',
+                confirmLabel: 'Delete'
+            }, async () => {
+                await SnapshotsApi.deleteSnapshot(snapshot.id)
+                const index = this.snapshots.indexOf(snapshot)
+                this.snapshots.splice(index, 1)
+                Alerts.emit('Successfully deleted snapshot.', 'confirmation')
+            })
+        },
+        isDevice: function (row) {
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/application/Snapshots.vue:178]: Discussed & I understand the better way now but not in scope at this time

Once Serban has implemented the first one and we establish a pattern, these will be retro'd upon working in the file

File: frontend/src/pages/application/Snapshots.vue [hunk 8]
```diff
@@ -135,6 +156,61 @@ export default {
...
+        },
+        showDownloadSnapshotDialog (snapshot) {
+            this.$refs.snapshotExportDialog.show(snapshot)
+        },
+        async downloadSnapshotPackage (snapshot) {
+            const ss = await SnapshotsApi.getFullSnapshot(snapshot.id)
+            const owner = snapshot.device || snapshot.project
+            const ownerType = snapshot.device ? 'device' : 'instance'
+            const packageJSON = {
+                name: `${owner.safeName || owner.name}`.replace(/[^a-zA-Z0-9-]/g, '-').toLowerCase(),
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/application/Snapshots.vue:156]: This was a direction dictated by the issue and PR to utelise end point.

At some future time, this indeed would be better suited to the backend however it is a duplicate of existing method

Out of scope.

File: frontend/src/pages/application/Snapshots.vue [hunk 9]
```diff
@@ -135,6 +156,61 @@ export default {
...
+                description: `${ownerType} snapshot, ${snapshot.name} - ${snapshot.description}`,
+                private: true,
+                version: '0.0.0-' + snapshot.id,
+                dependencies: ss.settings?.modules || {}
+            }
+            const element = document.createElement('a')
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/application/Snapshots.vue]: resolved in [15001ab](https://github.com/FlowFuse/flowfuse/pull/3885/commits/15001ab22c59d4581f2116b97633c5c418756099)

File: frontend/src/pages/application/Snapshots.vue [hunk 10]
```diff
@@ -123,6 +130,20 @@ export default {
             this.loading = true
             const data = await ApplicationApi.getSnapshots(this.application.id, null, null, null)
             this.snapshots = [...data.snapshots]
+            // For any snapshots that have no user and match the autoSnapshot name format
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply frontend/src/pages/application/Snapshots.vue]: This was a direction dictated by the issue and PR to utelise end point.

At some future time, this indeed would be better suited to the backend however it is a duplicate of existing method

Out of scope.

File: frontend/src/pages/instance/Snapshots/index.vue (hunk 4 above)
Comment (Steve-Mcl) [reply frontend/src/pages/instance/Snapshots/index.vue]: resolved in [30d8d50](https://github.com/FlowFuse/flowfuse/pull/3885/commits/30d8d50572972a1d4486fd848882f640385db857)

**Comment by Steve-Mcl** [comment]
@cstns all suggestions we discussed now in place and working and tests passing again. Ready for (re)review.

File: frontend/src/pages/application/Snapshots.vue [hunk 11]
```diff
@@ -135,6 +156,61 @@ export default {
...
+        },
+        // snapshot actions - delete
+        showDeleteSnapshotDialog (snapshot) {
+            Dialog.show({
+                header: 'Delete Snapshot',
+                text: 'Are you sure you want to delete this snapshot?',
+                kind: 'danger',
+                confirmLabel: 'Delete'
+            }, async () => {
+                await SnapshotsApi.deleteSnapshot(snapshot.id)
```
[the remark this answers is not in the document]
Comment (cstns) [reply frontend/src/pages/application/Snapshots.vue:172]: Was wondering if we should threat the possibility for the deleteSnapshot endpoint to fail

**Review by cstns — APPROVED** [review]
Looks good on my end! We can move forward even without the modal validation when deleting snapshots because i couldn't find any other delete dialogs that implement such a catch

**Pull request merged**